editor/osm_editor.cpp

```diff
@@ -646,8 +646,9 @@ void Editor::UploadChanges(string const & oauthToken, ChangesetTags tags,
               }
               else
               {
-                LOG(LDEBUG, ("Create case: uploading patched feature", osmFeature));
-                changeset.Modify(osmFeature);
+                LOG(LDEBUG, ("Create case: uploading new feature", osmFeature));
+                // There is another node nearby, but it is saver to upload a new node (#2298).
+                changeset.Create(feature);
```

Comment:
As was discussed in the issue, this implementation will always create the feature, even if it already exists. A better approach would be to try to compare features by some criteria and decide about modifying or creating based on it:

```
if (IsTheSameFeature(osmFeature, osmFeatureCopy)
{
  LOG(LDEBUG, ("Create case: uploading patched feature", osmFeature));
  changeset.Modify(osmFeature);
}
else
{
  LOG(LDEBUG, ("Create case: uploading new feature", osmFeature));
  // There is another feature node nearby, but it is safe to upload a new node (#2298).
  changeset.Create(feature);
}
```

`IsTheSameFeature` function can be implemented in different ways. It can be thoroughly tested with different combinations of tags. It would be great to debug and dump real tag values for some most popular OM types that are edited to write these tests properly.

Is there a way to calculate how many real duplicates were replaced with the existing code, and how many wrong types have been replaced?

Reply:
> Is there a way to calculate how many real duplicates were replaced with the existing code, and how many wrong types have been replaced?

I don't know how to prove this with numbers as this bug is very hard to spot. The changeset comments of OM "Updated ..." are very generic.

But when reviewing OM changesets I spot this bug ocasionally and I only remember one occasion where merging was a good thing to do. That was when a newly opened place got merged with a `disused:shop`. But even there the `disused:shop` tag remained in place, leading to inconsistent data.

Important to note is that the current merge decision is **nonsense**. The implementation just checks whether there is **any** feature nearby and not whether these features are **the same** before merging / overwriting. This algorithm merges a waste_basket into a fire_hydrant in case one is close by although this violates the ["one feature, one OSM element"](https://wiki.openstreetmap.org/wiki/One_feature,_one_OSM_element) rule of OSM.

Currently OM is trying to be smart, but it isn't. It's just damaging  OSM data. With this PR OM will always add a new place whenever the user clicks on "Add a place". If this leads to a duplicate this is an error type that is way easier to spot and fix in OSM. Moreover it is more "acceptable" to have these types of errors as OM is an offline editor, and editing based on up to two month old map data almost inevitably leads to some duplicates.

With this PR the overwhelming majority of cases will be handled correctly. An `IsTheSameFeature` check would be cool to have, but it won't prevent all duplicates and is difficult to implement. How can we tell that features are the same? If we use feature type and `name`, duplicates will be overlooked if the `name` is written slightly different. And just using feature type is not save e.g. if there are two clothes shops on top of each other (on different levels) in a shopping centre. In my opinion implementing this is not worth the effort at the moment as OM has more important issues.

This PR significantly reduces the number of wrong edits. That's why I suggest to merge this PR as a first step and open a new Issue for the merge algorithm.

Reply:
1. Thanks for the clarification! I agree that this should be an improvement. Is it possible to add some specific comment for these changesets, so we can later collect a database of all real cases and see where/how we can implement a better feature merge later?

2. Does it make sense to also create a note?

3. Btw, is there any external API/web service that can help to merge features more efficiently (and will be easier to support/extend than doing it in a client code)? Does it make sense to create such an API?

Reply:
> 1. 

That's a very good Idea. I added a "info:feature_close_by=yes" flag to the changeset meta data. This way one can easily filter for these changesets in e.g. OSMCha.

> 2.

No, notes are for human-human interaction only. Automatically adding them is not allowed.

> 3.

I don't know any API